### DISCUSSION TITLE
Tracking Java Native Memory With JDK Flight Recorder - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#312] [Announcements]
# Tracking Java Native Memory With JDK Flight Recorder - Gunnar Morling

Update Dec 18: This post is discussed on Hacker News 🍊

As regular readers of this blog will now, JDK Flight Recorder (JFR) is one of my favorite tools of the Java platform.
This low-overhead event recording engine built into the JVM is invaluable for observing the runtime characteristics of Java applications and identifying any potential performance issues.
JFR continues to become better and better with every new release,
with one recent addition being support for native memory tracking (NMT).

https://www.morling.dev/blog/tracking-java-native-memory-with-jdk-flight-recorder/

**Comment by ProAlex11** [comment]
the one thing which NMT does not report, despite what the name might suggest, is any memory allocated by native libraries, for instance invoked via JNI
If you're using glibc, then malloc does have information about that, and provides ways to read it, so it's a shame this isn't exposed. It would be quite helpful in the face of suspected native library memory leaks.

**Reply by gunnarmorling** [reply]
Yepp. jemalloc should help with that too, I haven't tried this myself yet, though. There also is [a proposal](https://mail.openjdk.org/pipermail/jdk-dev/2023-December/008491.html) by @tstuefe to add this capability to OpenJDK.

**Reply by tstuefe** [reply]
> the one thing which NMT does not report, despite what the name might suggest, is any memory allocated by native libraries, for instance invoked via JNI

With a noteworthy exception. Memory allocated via `Unsafe.allocateMemory` is tracked by NMT (shows up as "Other"). That takes care of e.g. netty native memory.

**Reply by gunnarmorling** [reply]
`Unsafe.allocateMemory` wouldn't be invoked though by native code, right? I.e. some Java code of Netty in that example.

**Reply by tstuefe** [reply]
Oh. Yes you are right. While they could be called from native code in theory, that is very uncommon and not straightforward (the function is not exported directly but via function pointer interface).
